### ISSUE TITLE
hidinput exception in Kivy. No such devices

### ISSUE THREAD

**Issue by RedSoxXx** [kivy/kivy#4406]
I'm using a Kivy Python framework in my project and USB touch panel for user interraction and I faced with problem. When my touch panel has lost connection (for a second) and then reconnect again the application get an exception HidinputException in hidinput.py file.

Here I show my stacktrace:

```
Exception in thread Thread-7:
 Traceback (most recent call last):
   File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
     self.run()
   File "/usr/lib/python2.7/threading.py", line 763, in run
     self.__target(*self.__args, **self.__kwargs)
   File "/usr/local/lib/python2.7/dist-packages/kivy/input/providers/hidinput.py", line 687, in _thread_run
     data = fd.read(struct_input_event_sz)
 IOError: [Errno 19] No such devices
```

**Comment by KeyWeeUsr** [comment]
Seems good to me. It's not provider's problem that your device disconnected. If it's only Thread exception, you can surely trace the line of code that checks for devices and when it occurs, run it manually (from your file, not from provider).

**Comment by RedSoxXx** [comment]
Sorry, but I'm not sure that I understand you correctly. Could you tell me what exactly  should I do ?

**Comment by KeyWeeUsr** [comment]
Line 687 in provider throws an error. If it's only when the device disconnects (which probably is), the import from provider a function that works with that line and call it manually if such an error happens. Chech how to handle thread errors [here](http://stackoverflow.com/q/2829329/5994041). Hmm... probably a welcome edit in provider could be that the thread would return some value for user to check.

**Comment by openedhardware** [comment]
Hi,
I have same problem with my touch screen.
I am using 10.1" touch screen from WaveShare.
http://www.waveshare.com/wiki/10.1inch_HDMI_LCD_(B)_(with_case)

Though I am using external power source for the touch screen, same errors are occasionally occurs and it is suffering me continuously.
After error is occurred, screen is locked up and does not accept any touch input.

Is there any way to recover the connection in the hidinput.py?
Do I need to modify it?

Thanks.

**Comment by dlernstrom** [comment]
I also face this problem when running a Raspberry Pi with a Kivy application as a display.  Occasionally, I'd like to plug in a USB mouse/keyboard in order to tweak the application.  Under these cases, I am forced to restart the entire Kivy Application, just so that I can use keyboard/mouse.

Strange thing is that Kivy itself is obviously reconnecting to the keyboard (despite the IOError crash above).  I know that because I'm still able to CTRL-C my way out of the Kivy Application.

**Comment by biank88** [comment]
I have the same problem whenever a usb keyboard (or any other devices) is unplugged from the Raspberry while a kivy app is running. In that case, `OSError: [Errno 19] No such device` is prompted in the log file and **the whole kivy app becomes extremely laggy**. If the usb device is reconnected I'm able to write characters but the application remains laggy until the app is restarted.

**Comment by KeyWeeUsr** [comment]
Closed by #5206

**Comment by jonryk** [comment]
Same issue on a RaspberryPi if I turn off the USB-bus for a connected HID.
Kivy "out of the box" doesn't properly handle system USB dis-/re-connections taking place while the app is running.

Everything works fine if Kivy runs under X/windows-system (since X then handles the HID and their dis-/re-connections, I expect), but it doesn't work for system-devices enabled in the .kivy/config.ini.

I guess this is why the issue has been overlooked / neglected: Everything works fine when running under X, and even if not under X; everything is fine as long as you don't disconnect / reconnect a USB-device while the app is running.

However, since I want to be able to disconnect and reconnect a USB HID while runing my Kivy-app on a stripped down raspbian without X, I'm out of luck...

The important issue is the fact that input from a device will not reach the app if the device is reconnected after the app has been started - So far I have not been able to find any solution (when not running X).
